RecoMTD/MeasurementDet/src/MTDDetLayerMeasurements.cc

```diff
@@ -0,0 +1,397 @@
+    if(enableME0Measurement)
+      {
+	checkME0RecHits(); 
+
+	// Create the chamber Id
+	ME0DetId detId(geoId.rawId());
+    
+	// Get the ME0-Segment which relies on this chamber
+	// Getting rechits right now, not segments - maybe it should be segments?
+	ME0SegmentCollection::range range = theME0RecHits->get(detId);
+
+	LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "Number of ME0 rechits available =  " << theME0RecHits->size()
+							   <<", from chamber: "<< detId<<std::endl;
+
+	// Create the MuonTransientTrackingRecHit
+	for (ME0SegmentCollection::const_iterator rechit = range.first; 
+	     rechit!=range.second; ++rechit){
+	  LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "On ME0 iteration " <<std::endl;
+	  result.push_back(MuonTransientTrackingRecHit::specificBuild(geomDet,&*rechit));
+	}
+	LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "Number of ME0 rechits = " << result.size()<<std::endl;
+      }
+  }
+*/
+/*
+// FIXME, these might be needed for debugging
+  else {
+    // wrong type
+    throw cms::Exception("MTDDetLayerMeasurements") << "The DetLayer with det " << geoId.det() << " subdet " << geoId.subdetId() << " is not a valid MTD DetLayer. ";
+    LogDebug("Track|RecoMTD|MTDDetLayerMeasurements")<< "The DetLayer with det " << geoId.det() << " subdet " << geoId.subdetId() << " is not a valid MTD DetLayer. ";
+  }
+*/
+  /*if (enableME0Measurement){
+    LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "(ME0): enabled"<<std::endl;
+  }
+
+  if (enableGEMMeasurement){
+    LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "(GEM): enabled"<<std::endl;
+    }*/
+  return result;
+}
+
+void MTDDetLayerMeasurements::checkMTDRecHits()
+{
+  LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "Checking MTD RecHits";
+  checkEvent();
+  auto cacheID = theEvent->cacheIdentifier();
+  if (cacheID == theMTDEventCacheID) return;
+
+  {
+    theEvent->getByToken(mtdToken_, theMTDRecHits);
+    theMTDEventCacheID = cacheID;
+  }
+  if(!theMTDRecHits.isValid())
+  {
+    throw cms::Exception("MTDDetLayerMeasurements") << "Cannot get MTD RecHits";
+    LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "Cannot get MTD RecHits";
+  }
+}
+
+///measurements method if already got the Event 
+MeasurementContainer
+MTDDetLayerMeasurements::measurements( const DetLayer* layer,
+				       const TrajectoryStateOnSurface& startingState,
+				       const Propagator& prop,
+				       const MeasurementEstimator& est) {
+  checkEvent();
+  return measurements(layer, startingState, prop, est, *theEvent);
+}
+
+
+MeasurementContainer
+MTDDetLayerMeasurements::measurements( const DetLayer* layer,
+				       const TrajectoryStateOnSurface& startingState,
+				       const Propagator& prop,
+				       const MeasurementEstimator& est,
+				       const edm::Event& iEvent) {
+  
+  MeasurementContainer result;  
+
+  std::vector<DetWithState> dss = layer->compatibleDets(startingState, prop, est);
```

Comment:
`const auto&` ?

Reply:
It returns a new vector of TSOS and pointers when you make this query.